{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Top protein functions for those that have > 10 taxonomic assignments\n",
    "* Date: July 21, 2019\n",
    "* Location on Bridges: /pylon5/mc5fr5p/hbagheri/09_Hadoop/protFuncFreq21July2019OUTPUT\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 1: Boa<sub>g</sub> query\n",
    "\n",
    "* Tax ids should be distinct\n",
    "\n",
    "```\n",
    "\n",
    "s: Sequence = input;\n",
    "protOut : output sum [string] of int;\n",
    "#taxCount : output collection [string] of int;\n",
    "#DefLinecount : output collection [string] of int;\n",
    "\n",
    "distinctTax := function (seq: Sequence): int{\n",
    "  taxSet : set of string;\n",
    "  foreach(i:int; def(seq.annotation[i]))\n",
    "   add(taxSet,seq.annotation[i].tax_id);\n",
    "   \n",
    "  return(len(taxSet));\n",
    "  \n",
    "};\n",
    "\n",
    "\n",
    "if (distinctTax(s) > 10){\n",
    " #taxCount [s.seqid] << distinctTax(s);\n",
    " #DefLinecount [s.seqid] << len(s.annotation);\n",
    "\n",
    " foreach(i:int; def(s.annotation[i]))\n",
    "   if (strfind(\"[\",s.annotation[i].defline)> 0)\n",
    "    protOut [trim(substring(s.annotation[i].defline, 0, strfind(\"[\",s.annotation[i].defline)))] << 1;\n",
    "}\n",
    "\n",
    "```\n",
    "\n",
    "* few lines of Output file, i.e. part-r-00000:\n",
    "\n",
    "```\n",
    "protOut[#3R-hydroxymyristoyl-] = 2\n",
    "protOut[&alpha-D-glucose-1-phosphatase] = 5\n",
    "protOut[&alpha-xylosidase] = 2\n",
    "protOut[&beta-D-glucoside glucohydrolase, periplasmic] = 4\n",
    "protOut[&gamma-glutamyl:cysteine ligase YbdK] = 2\n",
    "protOut[&gamma-glutamylputrescine synthetase] = 3\n",
    "protOut['2,4-dienoyl-CoA reductase (NADPH)] = 1\n",
    "protOut['Cold-shock' DNA-binding domain protein (plasmid)] = 1\n",
    "protOut['Cold-shock' DNA-binding domain protein] = 3355\n",
    "protOut['Cold-shock' DNA-binding domain, putative] = 4\n",
    "protOut['Cold-shock' DNA-binding domain-containing protein] = 2\n",
    "protOut['Cold-shock' DNA-binding domain] = 3\n",
    "protOut['Not a Proline racemase, nor 4-hydroxyproline epimerase (missing catalytic residues)] = 1\n",
    "protOut['Paired box' domain protein] = 5\n",
    "protOut['Ppf3p'] = 1\n",
    "protOut['Pyruvate oxidase (ubiquinone, cytochrome)] = 1\n",
    "protOut['Sequence is part of a 1.65-kb fragment of the lactose plasmid from Lactococcus lactis subsp cremoris SK11 that facilitates integration when plasmid is in L.lactis subsp. lactis.'; ORF] = 1\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 2: Post Processing\n",
    "\n",
    "* put  the last column to the beginning of each line:  \n",
    "```awk '{print $NF, $0}' part-r-00000 > part-r-00000_modified```\n",
    "\n",
    "\n",
    "* sort based on the first column in the reversed order:\n",
    "``` sort -nrk1 part-r-00000_modified > part-r-00000_modified_sorted_r ```\n",
    "\n",
    "\n",
    "* remove the first column:\n",
    "``` cut -d ' ' -f 2- part-r-00000_modified_sorted_r  > part-r-00000_modified_sorted_r_cut1 ```\n",
    "\n",
    "\n",
    "*  few lines of output\n",
    "\n",
    "```\n",
    "protOut[Uncharacterised protein] = 3476480\n",
    "protOut[transcriptional regulator] = 2044629\n",
    "protOut[membrane protein] = 1629015\n",
    "protOut[LysR family transcriptional regulator] = 871201\n",
    "protOut[ABC transporter ATP-binding protein] = 819880\n",
    "protOut[transposase] = 658213\n",
    "protOut[MFS transporter] = 652282\n",
    "protOut[oxidoreductase] = 633541\n",
    "protOut[cytochrome oxidase subunit 1, partial (mitochondrion)] = 559708\n",
    "protOut[ABC transporter permease] = 551235\n",
    "protOut[conserved hypothetical protein] = 486938\n",
    "protOut[lipoprotein] = 485403\n",
    "protOut[AraC family transcriptional regulator] = 392415\n",
    "protOut[TetR family transcriptional regulator] = 391287\n",
    "protOut[DNA-binding response regulator] = 380999\n",
    "protOut[GntR family transcriptional regulator] = 377254\n",
    "protOut[phage protein] = 365918\n",
    "protOut[transporter] = 356273\n",
    "protOut[DNA-binding protein] = 322177\n",
    "protOut[transmembrane protein] = 319394\n",
    "protOut[hydrolase] = 315531\n",
    "protOut[acetyltransferase] = 293665\n",
    "protOut[integral membrane protein] = 293034\n",
    "```\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## Top protein annotations <span style=\"color:red\"> 4 types?? </span>\n",
    "\n",
    "### Highly conserved\n",
    "* link: https://en.wikipedia.org/wiki/Conserved_sequence\n",
    "* Examples of highly conserved sequences include the RNA components of ribosomes present in all domains of life, the homeobox sequences widespread amongst Eukaryotes, and the tmRNA in Bacteria.\n",
    "\n",
    "### Highly mobile: Virus, plasmid  like antibiotic resistance\n",
    "\n",
    "### Highly generic\n",
    "\n",
    "\n",
    "## Top Annotations\n",
    "*  [<span style=\"color:blue\"> Uniport categories</span>](https://www.uniprot.org/statistics/Swiss-Prot)  \n",
    "* <span style=\"color:red\"> TODO: what categories from here? </span>\n",
    "\n",
    "##### membrane protein\n",
    "* ref: https://www.sciencedirect.com/science/article/pii/S0022283614005130\n",
    "* \"they underlie virtually all physiological processes in cells including key metabolic pathways, such as the respiratory chain and the photosystems, as well as the transport of solutes and signals across membranes.\"\n",
    "\n",
    "##### LysR family transcriptional regulator\n",
    "* https://www.annualreviews.org/doi/pdf/10.1146/annurev.mi.47.100193.003121\n",
    "*\n",
    "\n",
    "##### ABC transporter ATP-binding protein]\n",
    "* ref: http://www.jlr.org/content/42/7/1007.short\n",
    "\n",
    "\n",
    "##### oxidoreductase\n",
    "*\n",
    "\n",
    "##### conserved hypothetical protein\n",
    "* https://academic.oup.com/nar/article/32/18/5452/998719\n",
    "* https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1891709/\n",
    "\n",
    "##### lipoprotein\n",
    "* https://www.nejm.org/doi/pdf/10.1056/NEJM198904203201607\n",
    "* \"LIPOPROTEIN lipase, a hydrolytic enzyme produced by many tissues, is rate-limiting for the removal of lipoprotein triglycerides from the circulation\"\n",
    "\n",
    "##### family transcriptional regulator\n",
    "* https://mic.microbiologyresearch.org/content/journal/micro/10.1099/mic.0.2008/022772-0?crawler=true\n",
    "* The LysR-type transcriptional regulator (LTTR) family is a well-characterized group of transcriptional regulators. They are highly conserved and ubiquitous amongst bacteria, with functional orthologues identified in archaea and eukaryotic organisms ( PeÃÅrez-Rueda & Collado-Vides, 2001 ; Sun & Klein, 2004 ; Stec et al., 2006 ).\n",
    "\n",
    "##### transporter\n",
    "\n",
    "#####  acetyltransferase\n",
    "\n",
    "##### Uncharacterized protein conserved in bacteria\n",
    "* ```protOut[Uncharacterized protein conserved in bacteria] = 376245\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "### Boa<sub>g</sub> query: number of clusters with the specific protein functions\n",
    "\n",
    "```\n",
    "s: Sequence = input;\n",
    "clstrCount: output sum [string][string] of int;\n",
    "\n",
    "pfuncs := {\n",
    "    \"transcriptional regulator\", \"membrane protein\",\"LysR family transcriptional regulator\",\"cytochrome oxidase subunit 1, partial (mitochondrion)\",\"transposase\",\"conserved hypothetical protein\",\"16S rRNA methyltransferase\",\"lipoprotein\", \"Uncharacterized protein conserved in bacteria\"\n",
    "};\n",
    "\n",
    "for (j := 0; j < len(pfuncs); j++) {\n",
    "\texists(i: int; match(pfuncs[j], s.annotation[i].defline)){\n",
    "    foreach(k:int; def(s.cluster[k])){\n",
    "      if(s.cluster[k].similarity==95)\n",
    "          \t clstrCount [pfuncs[j]][s.cluster[k].cid] << 1;\n",
    "\n",
    "    }   \n",
    "  }  \n",
    "}\n",
    "```\n",
    "\n",
    "\n",
    "* Boa<sub>g</sub> file name on XSEDE: protFuncClstrJun132019.boa\n",
    "* Output:\n",
    "\n",
    "* Location on Bridges: ```/pylon5/mc5fr5p/hbagheri/09_Hadoop/protFuncClstrJun132019 ```\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#### rRNA protein annotations\n",
    "\n",
    "* use the output of the previous step: ```grep \"rRNA\" part-r-00000_modified_sorted_r_cut1 > rRNA_annotations```\n",
    "\n",
    "* sample output:\n",
    "\n",
    "  ```\n",
    "    protOut[16S rRNA methyltransferase] = 45667\n",
    "    protOut[23S rRNA methyltransferase] = 40755\n",
    "    protOut[23S rRNA (pseudouridine(1915)-N(3))-methyltransferase RlmH] = 20202\n",
    "    protOut[23S rRNA pseudouridylate synthase] = 20013\n",
    "    protOut[23S rRNA (guanosine(2251)-2'-O)-methyltransferase RlmB] = 18832\n",
    "    protOut[tRNA/rRNA methyltransferase] = 17298\n",
    "    protOut[rRNA methyltransferase] = 16836\n",
    "    protOut[16S rRNA (uracil(1498)-N(3))-methyltransferase] = 15866\n",
    "    protOut[23S rRNA (uracil-5-)-methyltransferase RumA] = 13934\n",
    "    protOut[rRNA large subunit methyltransferase] = 13582\n",
    "    protOut[16S rRNA (guanine(966)-N(2))-methyltransferase RsmD] = 13431\n",
    "    protOut[16S rRNA methyltransferase GidB] = 13199\n",
    "    protOut[16S rRNA-processing protein RimM] = 13093\n",
    "    protOut[16S rRNA pseudouridine(516) synthase] = 12613\n",
    "    protOut[rRNA methylase] = 11955\n",
    "    protOut[16S rRNA (cytidine(1402)-2'-O)-methyltransferase] = 11084\n",
    "    protOut[16S rRNA processing protein RimM] = 10909\n",
    "    protOut[rRNA maturation RNase YbeY] = 10063\n",
    "    protOut[23S rRNA pseudouridine synthase F] = 9916\n",
    "    protOut[23S rRNA (uracil(1939)-C(5))-methyltransferase RlmD] = 9690\n",
    "\n",
    "   ```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Final result based on the 2 previous Boa<sub>g</sub> query:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "|Category|Protein function|#of appearance|#of clusters |Reference\n",
    "|--|--| -- | -- | --|\n",
    "|Highly Conserved |transcriptional regulator |2515980 | |ref\n",
    "|                 |membrane protein         |2032297 |1194598 |ref\n",
    "| |LysR family transcriptional regulator | 1268891 |301611 | ref\n",
    "| |cytochrome oxidase subunit 1, partial (mitochondrion) | 1165451 | | ref\n",
    "| | transposase| 936112 | 340654| ref\n",
    "|Highly Conserved |conserved hypothetical protein | 668349 |792297 | ref\n",
    "| Highly Conserved|16S rRNA methyltransferase|58951|14639 |ref\n",
    "|Highly generic |lipoprotein |607562| 274952| ref\n",
    "|Highly Conserved|Uncharacterized protein conserved in bacteria|376245|29004 | ref\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
